File: microservices/auth-service/tracing/tracing.go
package tracing

import (
	"context"
	"log"
	"os"

	"go.opentelemetry.io/otel"
	"go.opentelemetry.io/otel/exporters/zipkin"
	"go.opentelemetry.io/otel/sdk/resource"
	tracesdk "go.opentelemetry.io/otel/sdk/trace"
	semconv "go.opentelemetry.io/otel/semconv/v1.27.0"
)

// InitializeTracer initializes and returns a TracerProvider with Zipkin exporter
func InitializeTracer(serviceName string) (*tracesdk.TracerProvider, error) {
	// Get Zipkin endpoint from environment or use default
	zipkinEndpoint := os.Getenv("OTEL_EXPORTER_ZIPKIN_ENDPOINT")
	if zipkinEndpoint == "" {
		zipkinEndpoint = "http://zipkin:9411/api/v2/spans"
	}

	// Create Zipkin exporter
	exporter, err := zipkin.New(zipkinEndpoint)
	if err != nil {
		return nil, err
	}

	// Create resource
	res, err := resource.New(context.Background(),
		resource.WithAttributes(
			semconv.ServiceNameKey.String(serviceName),
			semconv.ServiceVersionKey.String("1.0.0"),
		),
	)
	if err != nil {
		return nil, err
	}

	// Create TracerProvider
	tp := tracesdk.NewTracerProvider(
		tracesdk.WithBatcher(exporter),
		tracesdk.WithResource(res),
	)

	// Set global TracerProvider
	otel.SetTracerProvider(tp)

	log.Printf("Tracing initialized for service: %s", serviceName)
	return tp, nil
}

// Shutdown gracefully shuts down the tracer provider
func Shutdown(ctx context.Context, tp *tracesdk.TracerProvider) error {
	return tp.Shutdown(ctx)
}

File: microservices/department-management/app/tracing.py
import os
import logging
from opentelemetry import trace
from opentelemetry.sdk.trace import TracerProvider
from opentelemetry.sdk.trace.export import BatchSpanProcessor
from opentelemetry.exporter.zipkin.json import ZipkinExporter
from opentelemetry.sdk.resources import SERVICE_NAME, Resource
from opentelemetry.instrumentation.fastapi import FastAPIInstrumentor
from opentelemetry.instrumentation.requests import RequestsInstrumentor


def initialize_tracer(service_name: str) -> TracerProvider:
    """Initialize OpenTelemetry tracing for FastAPI service"""
    
    # Get Zipkin endpoint from environment or use default
    zipkin_endpoint = os.getenv(
        "OTEL_EXPORTER_ZIPKIN_ENDPOINT",
        "http://zipkin:9411/api/v2/spans"
    )
    
    # Create Zipkin exporter
    zipkin_exporter = ZipkinExporter(
        endpoint=zipkin_endpoint,
    )
    
    # Create resource
    resource = Resource(attributes={
        SERVICE_NAME: service_name,
        "service.version": "1.0.0",
    })
    
    # Create TracerProvider
    trace_provider = TracerProvider(resource=resource)
    trace_provider.add_span_processor(BatchSpanProcessor(zipkin_exporter))
    
    # Set global TracerProvider
    trace.set_tracer_provider(trace_provider)
    
    logging.info(f"Tracing initialized for service: {service_name}")
    return trace_provider


def instrument_app(app):
    """Instrument FastAPI app and requests library"""
    FastAPIInstrumentor.instrument_app(app)
    RequestsInstrumentor().instrument()


def shutdown_tracer(trace_provider: TracerProvider):
    """Gracefully shutdown tracer"""
    trace_provider.force_flush()
